### ISSUE TITLE
Convert model from .pth to .onnx in mmdeploy, the output image of onnxruntime and pytorch both perform badly

### ISSUE THREAD

**Issue by lbq779660843** [open-mmlab/mmpose#2634]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmpose/issues) and [Discussions](https://github.com/open-mmlab/mmpose/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmpose).

### Environment

OrderedDict([('sys.platform', 'linux'), ('Python', '3.8.10 (default, May 26 2023, 14:05:08) [GCC 9.4.0]'), ('CUDA available', True), ('numpy_random_seed', 2147483648), ('GPU 0', 'NVIDIA GeForce RTX 4070 Ti'), ('CUDA_HOME', '/usr/local/cuda'), ('NVCC', 'Cuda compilation tools, release 12.1, V12.1.66'), ('GCC', 'x86_64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0'), ('PyTorch', '2.0.0+cu118'), ('PyTorch compiling details', 'PyTorch built with:\n  - GCC 9.3\n  - C++ Version: 201703\n  - Intel(R) oneAPI Math Kernel Library Version 2022.2-Product Build 20220804 for Intel(R) 64 architecture applications\n  - Intel(R) MKL-DNN v2.7.3 (Git Hash 6dbeffbae1f23cbbeae17adb7b5b13f1f37c080e)\n  - OpenMP 201511 (a.k.a. OpenMP 4.5)\n  - LAPACK is enabled (usually provided by MKL)\n  - NNPACK is enabled\n  - CPU capability usage: AVX2\n  - CUDA Runtime 11.8\n  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_90,code=sm_90\n  - CuDNN 8.8.1  (built against CUDA 12.0)\n    - Built with CuDNN 8.7\n  - Magma 2.6.1\n  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.8, CUDNN_VERSION=8.7.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -D_GLIBCXX_USE_CXX11_ABI=0 -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_DISABLE_GPU_ASSERTS=ON, TORCH_VERSION=2.0.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=1, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, \n'), ('TorchVision', '0.15.1+cu118'), ('OpenCV', '4.8.0'), ('MMEngine', '0.8.0'), ('MMPose', '1.0.0+f78af77')])

comm                          0.1.3
command-not-found             0.3
diagnostic-common-diagnostics 1.11.0
mmcv                          2.0.0
mmdeploy                      1.2.0
mmdeploy-runtime              1.2.0
mmdeploy-runtime-gpu          1.2.0
mmdet                         3.0.0
mmdet3d                       1.1.0                /home/lbq/Codes/Python/mmdetection3d
mmengine                      0.8.0
mmpose                        1.0.0                /home/lbq/Codes/Python/mmpose
mmpretrain                    1.0.0rc7             /home/lbq/Codes/Python/mmpretrain
mmsegmentation                1.0.0
qt-gui-py-common              0.4.2
rqt_py_common                 0.5.3
snowballstemmer               2.0.0
ubuntu-drivers-common         0.0.0

### Reproduces the problem - code sample

'''  configs/body_2d_keypoint/rtmpose/coco/rtmpose-s_8xb256-420e_coco-slot-256x192.py  '''
_base_ = ['../../../_base_/default_runtime.py']

max_epochs = 600
stage2_num_epochs = 30
base_lr = 4e-3

train_cfg = dict(max_epochs=max_epochs, val_interval=60)
randomness = dict(seed=21)

optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='AdamW', lr=base_lr, weight_decay=0.),
    paramwise_cfg=dict(
        norm_decay_mult=0, bias_decay_mult=0, bypass_duplicate=True))

param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=1.0e-5,
        by_epoch=False,
        begin=0,
        end=1000),
    dict(
        # use cosine lr from 210 to 420 epoch
        type='CosineAnnealingLR',
        eta_min=base_lr * 0.05,
        begin=max_epochs // 2,
        end=max_epochs,
        T_max=max_epochs // 2,
        by_epoch=True,
        convert_to_iter_based=True),
]

auto_scale_lr = dict(base_batch_size=1024)

codec = dict(
    type='SimCCLabel',
    input_size=(192, 256),
    sigma=(4.9, 5.66),
    simcc_split_ratio=2.0,
    normalize=False,
    use_dark=False)

model = dict(
    type='TopdownPoseEstimator',
    data_preprocessor=dict(
        type='PoseDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True),
    backbone=dict(
        _scope_='mmdet',
        type='CSPNeXt',
        arch='P5',
        expand_ratio=0.5,
        deepen_factor=0.33,
        widen_factor=0.5,
        out_indices=(4, ),
        channel_attention=True,
        #norm_cfg=dict(type='SyncBN'),
        act_cfg=dict(type='SiLU'),
        init_cfg=dict(
            type='Pretrained',
            prefix='backbone.',
            checkpoint='https://download.openmmlab.com/mmpose/v1/projects/'
            'rtmpose/cspnext-s_udp-aic-coco_210e-256x192-92f5a029_20230130.pth'  # noqa
        )),
    head=dict(
        type='RTMCCHead',
        in_channels=512,
        out_channels=4,
        input_size=codec['input_size'],
        in_featuremap_size=(6, 8),
        simcc_split_ratio=codec['simcc_split_ratio'],
        final_layer_kernel_size=7,
        gau_cfg=dict(
            hidden_dims=256,
            s=128,
            expansion_factor=2,
            dropout_rate=0.,
            drop_path=0.,
            act_fn='SiLU',
            use_rel_bias=False,
            pos_enc=False),
        loss=dict(
            type='KLDiscretLoss',
            use_target_weight=True,
            beta=10.,
            label_softmax=True),
        decoder=codec),
    test_cfg=dict(flip_test=True))

data_root = 'data/coco_slot_2/'
dataset_type = 'SlotCocoDataset'
data_mode = 'topdown'

backend_args = dict(backend='local')

train_pipeline = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='RandomFlip', direction='horizontal'),
    dict(type='RandomHalfBody'),
    dict(
        type='RandomBBoxTransform', scale_factor=[0.6, 1.4], rotate_factor=80),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='mmdet.YOLOXHSVRandomAug'),
    dict(
        type='Albumentation',
        transforms=[
            dict(type='Blur', p=0.1),
            dict(type='MedianBlur', p=0.1),
            dict(
                type='CoarseDropout',
                max_holes=1,
                max_height=0.4,
                max_width=0.4,
                min_holes=1,
                min_height=0.2,
                min_width=0.2,
                p=1.),
        ]),
    dict(type='GenerateTarget', encoder=codec),
    dict(type='PackPoseInputs')
]
val_pipeline = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='PackPoseInputs')
]

train_pipeline_stage2 = [
    dict(type='LoadImage', backend_args=backend_args),
    dict(type='GetBBoxCenterScale'),
    dict(type='RandomFlip', direction='horizontal'),
    dict(type='RandomHalfBody'),
    dict(
        type='RandomBBoxTransform',
        shift_factor=0.,
        scale_factor=[0.75, 1.25],
        rotate_factor=60),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='mmdet.YOLOXHSVRandomAug'),
    dict(
        type='Albumentation',
        transforms=[
            dict(type='Blur', p=0.1),
            dict(type='MedianBlur', p=0.1),
            dict(
                type='CoarseDropout',
                max_holes=1,
                max_height=0.4,
                max_width=0.4,
                min_holes=1,
                min_height=0.2,
                min_width=0.2,
                p=0.5),
        ]),
    dict(type='GenerateTarget', encoder=codec),
    dict(type='PackPoseInputs')
]

train_dataloader = dict(
    batch_size=128,
    num_workers=2,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/train.json',
        data_prefix=dict(img='data/'),
        pipeline=train_pipeline,
    ))
val_dataloader = dict(
    batch_size=64,
    num_workers=2,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False, round_up=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/val.json',
        data_prefix=dict(img='data/'),
        test_mode=True,
        pipeline=val_pipeline,
    ))
test_dataloader = val_dataloader

default_hooks = dict(
    checkpoint=dict(save_best='PCK', rule='greater', max_keep_ckpts=1))

custom_hooks = [
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0002,
        update_buffers=True,
        priority=49),
    dict(
        type='mmdet.PipelineSwitchHook',
        switch_epoch=max_epochs - stage2_num_epochs,
        switch_pipeline=train_pipeline_stage2)
]

val_evaluator = [
    dict(type='PCKAccuracy', thr=0.2),
    dict(type='AUC'),
    dict(type='EPE'),
    #dict(type='NME', norm_mode='use_norm_item'),
    #dict(ann_file=data_root + 'annotations/val.json'),
]
test_evaluator = val_evaluator

### Reproduces the problem - command or script

''' convert script  '''
python3 tools/deploy.py \
    configs/mmpose/pose-detection_simcc_onnxruntime_dynamic.py \
    ../mmpose/configs/body_2d_keypoint/rtmpose/coco/rtmpose-s_8xb256-420e_coco-slot-256x192.py \
    ../mmpose/work_dirs/rtmpose-s_8xb256-420e_coco-slot-256x192/best_PCK_epoch_240.pth \
    demo/resources/slot_dark.jpg \
    --work-dir work_dir/rtmpose-ort-slot-dark/rtmpose-m \
    --device cpu \
    --show \
    --dump-info
    
''' test/evaluate script  ''' 
   python3 tools/test.py configs/body_2d_keypoint/rtmpose/coco/rtmpose-s_8xb256-420e_coco-slot-256x192.py \
    work_dirs/rtmpose-s_8xb256-420e_coco-slot-256x192/best_PCK_epoch_240.pth \
    --show-dir work_dirs/rtmpose-s_8xb256-420e_coco-slot-256x192/test_out  \
    --work-dir work_dirs/rtmpose-s_8xb256-420e_coco-slot-256x192/eval_out

### Reproduces the problem - error message

None

### Additional information

1. I convert my model which was trained by my custom dataset from .pth to .onnx in **mmdeploy**, the output image of onnxruntime and PyTorch both perform badly, but the .pth model can get normal results while I use the test/evaluate script above.

output_onnxruntime
![output_onnxruntime](https://github.com/open-mmlab/mmpose/assets/27773249/a1514a45-784a-4abc-a9e6-78d79aa0b421)
output_pytorch
![output_pytorch](https://github.com/open-mmlab/mmpose/assets/27773249/cd449d0d-7037-4db4-be27-1f281a867894)

normal results
![20181002000013-00-00 MP4_1457_0](https://github.com/open-mmlab/mmpose/assets/27773249/7737c97c-eb58-48bc-9ce7-ff52aea0b474)
![20181002000013-00-00 MP4_1457_3](https://github.com/open-mmlab/mmpose/assets/27773249/66773585-d218-4362-b957-39c215d430ee)
![20181002000013-00-00 MP4_1457_1](https://github.com/open-mmlab/mmpose/assets/27773249/2ec23df2-f3b3-420c-a702-487b7069b500)
![20181002000013-00-00 MP4_1457_2](https://github.com/open-mmlab/mmpose/assets/27773249/eb45ba53-6ea9-4cee-a039-91173e924cad)

“”“
2. I am using a custom dataset, which is strictly defined according to the coco dataset pose estimation standard。 
3. I get no idea.

**Comment by Tau-J** [comment]
Hi @lbq779660843 , thanks for using MMPose. Please note that RTMPose is a top-down algorithm, which means that you have to conduct object detection first to obtain the bbox of the target, then use rtmpose model to estimate the keypoint. It seems that you are directly passing the image into pose model, so it is normal to get wrong results.

**Comment by lbq779660843** [comment]
@Tau-J Surely, I'm directly passing a test image while trying to convert the RTMPose model from .pth to .onnx, so I feel confused about the difference in the generated results between the one from the test/evaluate script in MMPose and the convert script in MMDeploy.
Additionally, I'm trying to deploy the models on onnxruntime and tensorrt based on this [link](https://github.com/open-mmlab/mmpose/tree/main/projects/rtmpose/examples/RTMPose-Deploy)supplied by RTMPose, and I've trained the object detection model in mmdetection and converted this model to .onnx and .engine formats, the bbox result is fine but the poses are wrong.
 

![test](https://github.com/open-mmlab/mmpose/assets/27773249/1f9f06e8-2118-4de2-af5b-a3a5fb7949dd)

**Comment by lbq779660843** [comment]
I find out the main reason is due to MMDet model havn't been trained adequatly.
